File: core/src/main/java/no/elg/infiniteBootleg/world/render/ClientWorldRender.java
package no.elg.infiniteBootleg.world.render;

import static no.elg.infiniteBootleg.world.Block.BLOCK_SIZE;
import static no.elg.infiniteBootleg.world.Chunk.CHUNK_TEXTURE_SIZE;
import static no.elg.infiniteBootleg.world.ChunkColumn.Companion.FeatureFlag.TOP_MOST_FLAG;
import static no.elg.infiniteBootleg.world.GlobalLockKt.BOX2D_LOCK;

import com.badlogic.gdx.graphics.OrthographicCamera;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.math.MathUtils;
import com.badlogic.gdx.math.Matrix4;
import com.badlogic.gdx.math.Rectangle;
import com.badlogic.gdx.math.Vector3;
import com.badlogic.gdx.physics.box2d.Box2DDebugRenderer;
import com.badlogic.gdx.utils.ObjectMap;
import com.badlogic.gdx.utils.OrderedMap;
import com.badlogic.gdx.utils.OrderedSet;
import no.elg.infiniteBootleg.Main;
import no.elg.infiniteBootleg.Settings;
import no.elg.infiniteBootleg.world.Block;
import no.elg.infiniteBootleg.world.Chunk;
import no.elg.infiniteBootleg.world.ClientWorld;
import no.elg.infiniteBootleg.world.Location;
import no.elg.infiniteBootleg.world.box2d.WorldBody;
import org.jetbrains.annotations.NotNull;

/**
 * @author Elg
 */
public class ClientWorldRender implements WorldRender {

  @NotNull public final ClientWorld world;
  @NotNull private final Rectangle viewBound;
  @NotNull private final ClientChunksInView chunksInView;
  @NotNull private final Matrix4 m4 = new Matrix4();
  @NotNull OrderedMap<Chunk, TextureRegion> draw;
  @NotNull private final EntityRenderer entityRenderer;
  @NotNull private final SpriteBatch batch;
  @NotNull private final OrthographicCamera camera;
  @NotNull private final ChunkRenderer chunkRenderer;
  @NotNull private final Box2DDebugRenderer box2DDebugRenderer;
  @NotNull private final DebugChunkRenderer chunkDebugRenderer;
  private float lastZoom;

  public ClientWorldRender(@NotNull ClientWorld world) {
    viewBound = new Rectangle();
    chunksInView = new ClientChunksInView();
    this.world = world;
    draw = new OrderedMap<>();
    draw.orderedKeys().ordered = false; // improve remove

    chunkRenderer = new ChunkRenderer(this);
    entityRenderer = new EntityRenderer(this);

    camera = new OrthographicCamera();
    camera.setToOrtho(false);

    camera.zoom = 1f;
    camera.position.x = 0;
    camera.position.y = 0;

    batch = new SpriteBatch();
    batch.setProjectionMatrix(camera.combined);

    chunkDebugRenderer = new DebugChunkRenderer(this);
    box2DDebugRenderer = new Box2DDebugRenderer(false, false, true, true, true, true);
  }

  @Override
  public void render() {
    batch.setProjectionMatrix(camera.combined);
    chunkRenderer.render(20);

    draw.clear();
    draw.ensureCapacity(chunksInView.getSize());

    WorldBody worldBody = world.getWorldBody();
    float worldOffsetX = worldBody.getWorldOffsetX() * BLOCK_SIZE;
    float worldOffsetY = worldBody.getWorldOffsetY() * BLOCK_SIZE;
    int verticalStart = chunksInView.getVerticalStart();
    int verticalEnd = chunksInView.getVerticalEnd();
    for (int chunkY = verticalStart; chunkY < verticalEnd; chunkY++) {
      int horizontalStart = chunksInView.getHorizontalStart();
      int horizontalEnd = chunksInView.getHorizontalEnd();
      for (int chunkX = horizontalStart; chunkX < horizontalEnd; chunkX++) {
        Chunk chunk = world.getChunk(chunkX, chunkY, false);
        if (chunk == null) {
          int loadingChunkX = chunkX;
          int loadingChunkY = chunkY;
          Main.inst()
              .getScheduler()
              .executeAsync(() -> world.loadChunk(loadingChunkX, loadingChunkY));
          continue;
        }
        chunk.view();

        // No need to update texture when out of view, but in loaded zone
        if (chunkY == verticalEnd - 1
            || chunkY == verticalStart
            || chunkX == horizontalStart
            || chunkX == horizontalEnd - 1) {
          continue;
        }

        //noinspection LibGDXFlushInsideLoop
        if (chunk.isAllAir()
            && chunk.getChunkColumn().isChunkAboveTopBlock(chunk.getChunkY(), TOP_MOST_FLAG)) {
          continue;
        }

        // get texture here to update last viewed in chunk
        //noinspection LibGDXFlushInsideLoop
        TextureRegion textureRegion = chunk.getTextureRegion();
        if (textureRegion == null) {
          continue;
        }
        draw.put(chunk, textureRegion);
      }
    }
    batch.begin();
    for (ObjectMap.Entry<Chunk, TextureRegion> entry : draw.entries()) {
      var dx = entry.key.getChunkX() * CHUNK_TEXTURE_SIZE + worldOffsetX;
      var dy = entry.key.getChunkY() * CHUNK_TEXTURE_SIZE + worldOffsetY;
      batch.draw(entry.value, dx, dy, CHUNK_TEXTURE_SIZE, CHUNK_TEXTURE_SIZE);
    }
    entityRenderer.render();
    batch.end();
    if (Settings.renderChunkBounds && Settings.debug) {
      chunkDebugRenderer.render();
    }
    if (Settings.renderBox2dDebug && Settings.debug) {
      synchronized (BOX2D_LOCK) {
        box2DDebugRenderer.render(worldBody.box2dWorld, m4);
      }
    }
  }

  @Override
  public void update() {
    camera.update();
    m4.set(camera.combined).scl(Block.BLOCK_SIZE);

    float width = camera.viewportWidth * camera.zoom;
    float height = camera.viewportHeight * camera.zoom;

    if (!getWorld().getWorldTicker().isPaused()) {

      WorldBody worldBody = world.getWorldBody();
      float worldOffsetX = worldBody.getWorldOffsetX() * BLOCK_SIZE;
      float worldOffsetY = worldBody.getWorldOffsetY() * BLOCK_SIZE;

      float w = width * Math.abs(camera.up.y) + height * Math.abs(camera.up.x);
      float h = height * Math.abs(camera.up.y) + width * Math.abs(camera.up.x);
      viewBound.set(
          (camera.position.x - worldOffsetX) - w / 2,
          (camera.position.y - worldOffsetY) - h / 2,
          w,
          h);

      chunksInView.setHorizontalStart(
          MathUtils.floor(viewBound.x / CHUNK_TEXTURE_SIZE) - CHUNKS_IN_VIEW_HORIZONTAL_PHYSICS);
      chunksInView.setHorizontalEnd(
          MathUtils.floor((viewBound.x + viewBound.width + CHUNK_TEXTURE_SIZE) / CHUNK_TEXTURE_SIZE)
              + CHUNKS_IN_VIEW_HORIZONTAL_PHYSICS);

      chunksInView.setVerticalStart(
          MathUtils.floor(viewBound.y / CHUNK_TEXTURE_SIZE) - CHUNKS_IN_VIEW_PADDING_RENDER);
      chunksInView.setVerticalEnd(
          MathUtils.floor(
                  (viewBound.y + viewBound.height + CHUNK_TEXTURE_SIZE) / CHUNK_TEXTURE_SIZE)
              + CHUNKS_IN_VIEW_PADDING_RENDER);

      if (Math.abs(lastZoom - camera.zoom) > SKYLIGHT_ZOOM_THRESHOLD) {
        lastZoom = camera.zoom;
      }
    }
  }

  @Override
  public void resize(int width, int height) {
    Vector3 old = camera.position.cpy();
    camera.setToOrtho(false, width, height);
    camera.position.set(old);
    update();
  }

  @Override
  public void dispose() {
    batch.dispose();
    chunkRenderer.dispose();
    box2DDebugRenderer.dispose();
    chunkDebugRenderer.dispose();
  }

  @Override
  public @NotNull ClientWorld getWorld() {
    return world;
  }

  /**
   * @return How many blocks there currently are horizontally on screen
   */
  public int blocksHorizontally() {
    return (int) Math.ceil(camera.viewportWidth * camera.zoom / BLOCK_SIZE) + 1;
  }

  @Override
  public boolean isOutOfView(@NotNull Chunk chunk) {
    return chunksInView.isOutOfView(chunk.getChunkX(), chunk.getChunkY());
  }

  @NotNull
  public ClientChunksInView getChunksInView() {
    return chunksInView;
  }

  public @NotNull OrthographicCamera getCamera() {
    return camera;
  }

  public @NotNull ChunkRenderer getChunkRenderer() {
    return chunkRenderer;
  }

  public @NotNull SpriteBatch getBatch() {
    return batch;
  }

  @NotNull
  @Override
  public OrderedSet<Location> getChunkLocationsInView() {
    return ChunksInView.Companion.toSet(chunksInView);
  }
}

File: core/src/main/kotlin/no/elg/infiniteBootleg/world/render/ClientWorldRender.kt
package no.elg.infiniteBootleg.world.render

import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.math.MathUtils
import com.badlogic.gdx.math.Matrix4
import com.badlogic.gdx.math.Rectangle
import com.badlogic.gdx.physics.box2d.Box2DDebugRenderer
import com.badlogic.gdx.scenes.scene2d.Stage
import com.badlogic.gdx.scenes.scene2d.utils.DragAndDrop
import com.badlogic.gdx.utils.Disposable
import io.github.oshai.kotlinlogging.KotlinLogging
import no.elg.infiniteBootleg.Settings
import no.elg.infiniteBootleg.api.Renderer
import no.elg.infiniteBootleg.api.render.OverlayRenderer.Companion.isInactive
import no.elg.infiniteBootleg.inventory.container.InterfaceId
import no.elg.infiniteBootleg.inventory.container.OwnedContainer
import no.elg.infiniteBootleg.inventory.ui.createContainerActor
import no.elg.infiniteBootleg.main.ClientMain
import no.elg.infiniteBootleg.screens.StageScreen
import no.elg.infiniteBootleg.util.ChunkCoord
import no.elg.infiniteBootleg.util.IBVisWindow
import no.elg.infiniteBootleg.util.WorldCompactLoc
import no.elg.infiniteBootleg.util.WorldCoordNumber
import no.elg.infiniteBootleg.util.decompactLocX
import no.elg.infiniteBootleg.util.decompactLocY
import no.elg.infiniteBootleg.util.safeUse
import no.elg.infiniteBootleg.world.BOX2D_LOCK
import no.elg.infiniteBootleg.world.blocks.Block
import no.elg.infiniteBootleg.world.chunks.Chunk
import no.elg.infiniteBootleg.world.managers.interfaces.InterfaceManager
import no.elg.infiniteBootleg.world.render.ChunksInView.Companion.chunkColumnsInView
import no.elg.infiniteBootleg.world.render.ChunksInView.Companion.iterator
import no.elg.infiniteBootleg.world.render.debug.AirBlockRenderer
import no.elg.infiniteBootleg.world.render.debug.BlockLightDebugRenderer
import no.elg.infiniteBootleg.world.render.debug.DebugChunkRenderer
import no.elg.infiniteBootleg.world.render.debug.EntityMarkerBlockRenderer
import no.elg.infiniteBootleg.world.render.debug.TopBlockChangeRenderer
import no.elg.infiniteBootleg.world.world.ClientWorld
import kotlin.math.abs

private val logger = KotlinLogging.logger {}

/**
 * @author Elg
 */
class ClientWorldRender(override val world: ClientWorld) : WorldRender {

  private val viewBound: Rectangle = Rectangle()
  private val box2dDebugM4 = Matrix4()
  private val renderers: Array<Renderer> = arrayOf(
    CachedChunkRenderer(this),
    AirBlockRenderer(this),
    HoveringBlockRenderer(this),
    EntityRenderer(this),
    DebugChunkRenderer(this),
    BlockLightDebugRenderer(this),
    TopBlockChangeRenderer(this),
    FuturePositionRenderer(this),
    EntityMarkerBlockRenderer(this)
  )

  private var lastZoom = 0f

  private val maybeStage: Stage?
    get() {
      return (ClientMain.inst().screen as? StageScreen)?.stage.also { stage ->
        if (stage == null) {
          logger.warn { "Could not get stage from screen ${ClientMain.inst().screen}" }
        }
      }
    }

  val chunksInView: ClientChunksInView = ClientChunksInView()
  val batch: SpriteBatch = SpriteBatch()
  val camera: OrthographicCamera = OrthographicCamera().also {
    it.setToOrtho(false)
    it.zoom = 1f
    it.position.x = 0f
    it.position.y = 0f
  }
  val chunkRenderer: ChunkRenderer = ChunkRenderer(this)
  val box2DDebugRenderer: Box2DDebugRenderer by lazy { Box2DDebugRenderer(true, false, false, false, true, false) }

  private val dad: DragAndDrop = DragAndDrop()

  val interfaceManager = InterfaceManager()

  fun openInterface(interfaceId: InterfaceId, createIfMissing: () -> IBVisWindow? = { null }) {
    interfaceManager.openInterface(interfaceId, maybeStage ?: return, createIfMissing)
  }

  fun closeInterface(interfaceId: InterfaceId) {
    interfaceManager.closeInterface(interfaceId)
  }

  fun toggleInterface(interfaceId: InterfaceId, createIfMissing: () -> IBVisWindow? = { null }) {
    interfaceManager.toggleInterface(interfaceId, maybeStage ?: return, createIfMissing)
  }

//  /**
//   * Register an IContainer with the UI, if one is already registered return a saved instance
//   *
//   * @return The ContainerActor for the given container
//   */
//  fun getContainerActor(ownedContainer: OwnedContainer): CompletableFuture<WindowAndStage>? {
//    val stage = maybeStage ?: return null
//    val interfaceId = ownedContainer.owner.toInterfaceId()
//
//    val storedActor = interfaceManager.getInterface(interfaceId)
//    return if (storedActor == null) {
//      Main.inst().scheduler.executeSync {
//        val actor = interfaceManager.getInterface(interfaceId) ?: createContainerActor(ownedContainer)
//        WindowAndStage(actor, stage)
//      }
//    } else {
//      CompletableFuture.completedFuture(WindowAndStage(storedActor, stage))
//    }
//  }

  fun createContainerActor(ownedContainer: OwnedContainer) = world.createContainerActor(ownedContainer, dad, batch)

  fun lookAt(loc: WorldCompactLoc) = lookAt(loc.decompactLocX(), loc.decompactLocY())
  fun lookAt(worldX: WorldCoordNumber, worldY: WorldCoordNumber) {
    camera.position.set(worldX.toFloat() * Block.BLOCK_SIZE, worldY.toFloat() * Block.BLOCK_SIZE, 0f)
    update()
  }

  override fun render() {
    batch.projectionMatrix = camera.combined
    chunkRenderer.renderMultiple()
    batch.safeUse {
      for (renderer in renderers) {
        if (renderer.isInactive) continue
        if (!batch.isDrawing) {
          logger.warn { "Batch was not drawing when rendering, starting batch" }
          batch.begin()
        }
        batch.projectionMatrix = camera.combined
        renderer.render()
      }
    }
    if (Settings.renderBox2dDebug) {
      synchronized(BOX2D_LOCK) {
        box2DDebugRenderer.render(world.worldBody.box2dWorld, box2dDebugM4)
      }
    }
  }

  override fun update() {
    camera.update()
    box2dDebugM4.set(camera.combined).scl(Block.BLOCK_SIZE.toFloat())
    val width = camera.viewportWidth * camera.zoom
    val height = camera.viewportHeight * camera.zoom
    val w = width * abs(camera.up.y) + height * abs(camera.up.x)
    val h = height * abs(camera.up.y) + width * abs(camera.up.x)
    viewBound[camera.position.x - w / 2, camera.position.y - h / 2, w] = h
    chunksInView.horizontalStart = MathUtils.floor(viewBound.x / Chunk.CHUNK_TEXTURE_SIZE) - WorldRender.CHUNKS_IN_VIEW_HORIZONTAL_PHYSICS
    chunksInView.horizontalEnd =
      (MathUtils.floor((viewBound.x + viewBound.width + Chunk.CHUNK_TEXTURE_SIZE) / Chunk.CHUNK_TEXTURE_SIZE) + WorldRender.CHUNKS_IN_VIEW_HORIZONTAL_PHYSICS)
    chunksInView.verticalStart = MathUtils.floor(viewBound.y / Chunk.CHUNK_TEXTURE_SIZE) - WorldRender.CHUNKS_IN_VIEW_PADDING_RENDER
    chunksInView.verticalEnd =
      (MathUtils.floor((viewBound.y + viewBound.height + Chunk.CHUNK_TEXTURE_SIZE) / Chunk.CHUNK_TEXTURE_SIZE) + WorldRender.CHUNKS_IN_VIEW_PADDING_RENDER)
    if (abs(lastZoom - camera.zoom) > WorldRender.SKYLIGHT_ZOOM_THRESHOLD) {
      lastZoom = camera.zoom
    }
  }

  override fun resize(width: Int, height: Int) {
    val old = camera.position.cpy()
    camera.setToOrtho(false, width.toFloat(), height.toFloat())
    camera.position.set(old)
    update()
  }

  override fun dispose() {
    batch.dispose()
    chunkRenderer.dispose()
    box2DDebugRenderer.dispose()
    for (renderer in renderers.filterIsInstance<Disposable>()) {
      renderer.dispose()
    }
  }

  override fun isOutOfView(chunk: Chunk): Boolean = chunksInView.isOutOfView(chunk.chunkX, chunk.chunkY)
  override fun isOutOfView(chunkX: ChunkCoord, chunkY: ChunkCoord): Boolean = chunksInView.isOutOfView(chunkX, chunkY)
  override fun isInView(chunkX: ChunkCoord, chunkY: ChunkCoord): Boolean = chunksInView.isInView(chunkX, chunkY)

  override val chunkLocationsInView get() = chunksInView.iterator()
  override val chunkColumnsInView get() = chunksInView.chunkColumnsInView()
}
